rfcs/GraphQLOverWebSocket.md

```diff
@@ -0,0 +1,191 @@
+
+## Communication
+
+The WebSocket sub-protocol for this specification is: `graphql-transport-ws`.
+
+Messages are represented through the JSON structure and are stringified before being sent over the network. They are bidirectional, meaning both the server and the client must conform to the specified message structure.
+
+**All** messages contain the `type` field outlining the action this message describes. Depending on the type, the message can contain two more _optional_ fields:
+
+- `id` used for uniquely identifying server responses and connecting them with the client's requests
+- `payload` holding the extra "payload" information to go with the specific message type
+
+The server can close the socket (kick the client off) at any time. The close event dispatched by the server is used to describe the fatal error to the client.
+
+The client closes the socket and the connection by dispatching a `1000: Normal Closure` close event to the server indicating a normal closure.
+
+## Message types
+
+### `ConnectionInit`
+
+Direction: **Client -> Server**
+
+Indicates that the client wants to establish a connection within the existing socket. This connection is **not** the actual WebSocket communication channel, but is rather a frame within it asking the server to allow future operation requests.
+
+The server must receive the connection initialisation message within the allowed waiting time specified in the `connectionInitWaitTimeout` parameter during the server setup. If the client does not request a connection within the allowed timeout, the server will close the socket with the event: `4408: Connection initialisation timeout`.
+
+If the server receives more than one `ConnectionInit` message at any given time, the server will close the socket with the event `4429: Too many initialisation requests`.
+
+```typescript
+interface ConnectionInitMessage {
+  type: 'connection_init';
+  payload?: Record<string, unknown>;
+}
+```
+
+### `ConnectionAck`
+
+Direction: **Server -> Client**
+
+Expected response to the `ConnectionInit` message from the client acknowledging a successful connection with the server.
+
+```typescript
+interface ConnectionAckMessage {
+  type: 'connection_ack';
+}
+```
+
+The client is now **ready** to request subscription operations.
+
+### `Subscribe`
+
+Direction: **Client -> Server**
+
+Requests an operation specified in the message `payload`. This message provides a unique ID field to connect published messages to the operation requested by this message.
+
+If there is already an active subscriber for a streaming operation matching the provided ID, the server will close the socket immediately with the event `4409: Subscriber for <unique-operation-id> already exists`. The server may not assert this rule for operations returning a single result as they do not require reservations for additional future events.
```

Comment:
1. Would closing the entire socket just because of a duplicated id too harsh? Why don't we simply send back a `type = error` message so the client can simply assign a new id and resend the subscribe operation?

2. "The server may not assert this rule for operations returning a single result as they do not require reservations for additional future events." This can create confusing cases. I think it is better to be strict and enforce this rule for both single and multiple-result operations.

Reply:
You should make sure never to have ID collisions. However, if you do have one on a streaming operation, besides it being an implementation bug - it could also be an attack. In either case, best way to stress the responsibility, and stay secure; is to close the socket immediately.

If you have a single emitted result, you're technically doing an old fashioned HTTP request. No harm can be done here, as the result is simply released bound with an ID. The Protocol, thus, leaves it to the user if he wants to implement this constraint or not.

Reply:
Following the excellent findings from https://github.com/enisdenjo/graphql-ws/issues/95 I have made further Protocol adaptations in https://github.com/graphql/graphql-over-http/pull/140/commits/edc4ab51ea206d2052cf108915d8e04ccdcabe0b.

The Protocol now enforces ID uniqueness across all operation types.